marshmallow/schema.py

```diff
@@ -714,6 +714,20 @@ def _init_fields(self):
             self._bind_field(field_name, field_obj)
             fields_dict[field_name] = field_obj
 
+        dump_data_keys = [
+            obj.data_key or name for name, obj in iteritems(fields_dict) if not obj.load_only
+        ]
+        if len(dump_data_keys) != len(set(dump_data_keys)):
+            data_keys_duplicates = {x for x in dump_data_keys if dump_data_keys.count(x) > 1}
+            raise ValueError('Duplicate data_keys: {}'.format(data_keys_duplicates))
+
+        load_attributes = [
+            obj.attribute or name for name, obj in iteritems(fields_dict) if not obj.dump_only
+        ]
+        if len(load_attributes) != len(set(load_attributes)):
+            attributes_duplicates = {x for x in load_attributes if load_attributes.count(x) > 1}
+            raise ValueError('Duplicate attributes: {}'.format(attributes_duplicates))
+
         return fields_dict
```

Comment:
I'm not sure what the code style/practice for the project is with regard to utility functions for _everything_, but I might consider putting the duplicate detection/selection into a separate function. things could be abbreviated to something like the following:

```python
duplicate_dump_data_keys = find_duplicates([
    obj.data_key or name for name, obj in iteritems(fields_dict) if not obj.load_only
])
if duplicate_dump_data_keys is not None:
    raise ValueError('Duplicate data_keys: {}'.format(duplicate_dump_data_keys))
```

Reply:
Yeah, I wondered what I could factorize and I figured factorizing just the set comprehension was not worth it.

But indeed, find_duplicates can also handle the len(list) != len(set) shortcut comparison (return None if same length) and the duplicate list generation (the set comprehension).

Thanks for reviewing.

Reply:
We could shorten the code by calculating the list of duplicates first, but I figured it would be better performance-wise to pass fast in the nominal no-dups case and only calculate the list if there are duplicates. This is a design choice/trade-off that makes the find_duplicates method a bit less generic. We could leave it inlined here. That's not a huge code duplication.

No strong feeling either way.

Reply:
I lean towards keeping this as is, if only for the performance consideration @lafrech pointed out. And the duplication is minimal.